include/wil/toolhelp32.h

```diff
@@ -0,0 +1,133 @@
+#include "resource.h"
+#ifndef __WIL_TOOLHELP32_INCLUDED
+#define __WIL_TOOLHELP32_INCLUDED
+#include <TlHelp32.h>
+#include <processthreadsapi.h>
+namespace wil
+{
+namespace details
+{
+    template <typename TEntry, typename TEnumApi, typename TCallback>
+    void do_enum_snapshot(HANDLE handle, TEntry& entry, TEnumApi&& enumApiFirst, TEnumApi&& enumApiNext, TCallback&& callback)
+    {
+        using result_t = decltype(callback(TEntry{}));
+        bool enumResult = enumApiFirst(handle, &entry);
+        if (!enumResult)
+            return;
+
+        do
+        {
+            if constexpr (wistd::is_void_v<result_t>)
+            {
+                callback(entry);
+            }
+            else if constexpr (wistd::is_same_v<result_t, bool>)
+            {
+                if (callback(entry))
+                    return;
+            }
+            else
+            {
+                static_assert(
+                    [] {
+                        return false;
+                    }(),
+                    "Callback must return void or bool");
+            }
+            enumResult = enumApiNext(handle, &entry);
+        } while (enumResult);
+    }
+} // namespace details
+
+template <typename TCallback>
+void for_each_process(TCallback&& callback)
+{
+    PROCESSENTRY32 entry{};
+    entry.dwSize = sizeof(entry);
+    details::do_enum_snapshot(
+        unique_handle{CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0)}.get(),
+        entry,
+        &Process32First,
+        &Process32Next,
+        wistd::forward<TCallback>(callback));
+}
+
+template <typename TCallback>
+void for_each_thread(TCallback&& callback)

```

Comment:
This name is confusing. On its own, I would expect it to enumerate the threads in the _current_ process, however that is not what this is doing.

Reply:
- `for_each_thread_global` for all system threads
- `for_each_thread_of_process(callback)` for current thread
- `for_each_thread_of_process(pid, callback)` for threads of specified process ?

Reply:
> Not creative enough for names

Me too... Only suggestions I can come up with at the moment are `for_each_system_thread` and `for_each_process_thread` that has an overload with a PID (otherwise assumes current process)